src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_Call.cs

```diff
@@ -364,11 +376,219 @@ private static bool IsSafeForReordering(BoundExpression expression, RefKind kind
             }
         }
 
+        private ImmutableArray<BoundExpression> VisitArguments(
+            SyntaxNode syntax,
+            ImmutableArray<BoundExpression> arguments,
+            Symbol methodOrIndexer,
+            ImmutableArray<int> argsToParamsOpt,
+            ref ImmutableArray<RefKind> argumentRefKindsOpt,
+            [NotNullIfNotNull("visitedReceiver")] ref BoundExpression? rewrittenReceiver,
+            out ArrayBuilder<LocalSymbol>? temps,
+            out BitVector positionsAssignedToTemp,
+            out bool receiverIsArgumentSideEffectSequence)
+        {
+            Debug.Assert(argumentRefKindsOpt.IsDefault || argumentRefKindsOpt.Length == arguments.Length);
+            var requiresInstanceReceiver = methodOrIndexer.RequiresInstanceReceiver() && methodOrIndexer is not MethodSymbol { MethodKind: MethodKind.Constructor };
+            Debug.Assert(!requiresInstanceReceiver || rewrittenReceiver != null || _inExpressionLambda);
+            temps = null;
+            positionsAssignedToTemp = BitVector.Null;
+            receiverIsArgumentSideEffectSequence = false;
+
+            if (arguments.IsEmpty)
+            {
+                return arguments;
+            }
+
+            var visitedArgumentsBuilder = ArrayBuilder<BoundExpression>.GetInstance(arguments.Length);
+            ArrayBuilder<BoundExpression>? sideEffectsBuilder = null;
+            int lastIndexStored = -1;
+            bool receiverStored = false;
+            ArrayBuilder<RefKind>? refKindsBuilder = null;
+            var parameters = methodOrIndexer.GetParameters();
+
+            for (int i = 0; i < arguments.Length; i++)
+            {
+                InterpolatedStringHandlerData? nullableInterpolatedData = setupInterpolationDataIfNeeded(
+                    i,
+                    argumentRefKindsOpt,
+                    ref receiverStored,
+                    ref lastIndexStored,
+                    ref rewrittenReceiver,
+                    ref temps,
+                    ref sideEffectsBuilder,
+                    ref positionsAssignedToTemp);
+
+                visitedArgumentsBuilder.Add(VisitExpression(arguments[i]));
+
+                if (nullableInterpolatedData is not null)
+                {
+                    var interpolationData = nullableInterpolatedData.GetValueOrDefault();
+                    foreach (var placeholder in interpolationData.ArgumentPlaceholders)
+                    {
+                        RemovePlaceholderReplacement(placeholder);
+                    }
+                }
+            }
+
+            if (temps != null)
+            {
+                Debug.Assert(temps.Count > 0);
+                Debug.Assert(sideEffectsBuilder!.Count > 0);
+                Debug.Assert(!positionsAssignedToTemp.IsNull);
+                Debug.Assert(receiverStored == requiresInstanceReceiver);
```

Comment:
>receiverStored == requiresInstanceReceiver [](http://example.com/codeflow?start=29&length=42)

This looks suspicious. We need to store the recevier and any other argument only if it used to create a builder.